### ISSUE TITLE
[Feature request] Parallel formatting

### ISSUE THREAD

**Issue by safesparrow** [fsprojects/fantomas#2624]
I propose we run formatting analysis in parallel, with a config option to control this.
The existing way of Fantomas deals with this problem is running most steps sequentially, I believe.

## Pros and Cons 

The advantages of making this adjustment to Fantomas are: much smaller wall-clock time of running fantomas.

The disadvantages of making this adjustment to Fantomas are:
- Some chance for race conditions
- Higher maximum CPU utilisation
- Higher maximum IO utilisation

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S?

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] I have read the [Contribution Guidelines](https://github.com/fsprojects/fantomas/blob/main/CONTRIBUTING.md).
* [x] I have [searched both open and closed suggestions on this site](https://github.com/fsprojects/fantomas/issues) and believe this is not a duplicate

Please tick all that apply:
* [x] This is not a breaking change to Fantomas
* [x] I or my company would be willing to help implement and/or test this
* [ ] This suggestion is part of the [Microsoft style guide](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting) (please add a link to section if so)
* [ ] This suggestion is part of the [G-Research style guide](https://github.com/G-Research/fsharp-formatting-conventions) (please add a link to section if so)

**Comment by nojaf** [comment]
Hello, the simplest thing to do would be to take a look at what can be done in parallel in the cli tool.
When formatting multiple files, that could be done in parallel. What happens inside the file (in `Fantomas.Core`) would be trickier and should be attempted in a later phase.

I'm all ears for a proposal 😉

**Comment by safesparrow** [comment]
In the CLI case just formatting all source files in parallel would give more than enough.

I don't know how slow formatting a single file can be in the tooling case, but I imagine it's quite fast, so I don't think there is a need to speed that up.

**Comment by safesparrow** [comment]
I attempted at PoC but for some reason a simple `Array.Parallel.map` for `processFile`/`stringToFile` is bottlenecked and doesn't use most cores on my machine (8C/16T). I don't yet understand what's happening:
<img width="1812" alt="image" src="https://user-images.githubusercontent.com/2478401/202059571-8d70f9b2-9fcc-4e8a-bcf2-33cd96eac43c.png">

**Comment by nojaf** [comment]
> I attempted at PoC but for some reason a simple Array.Parallel.map for processFile/stringToFile is bottlenecked 

That rings a bell, I remember adding in an `Async.Parallel` at some point but it was worse than before. I'm glad you found a fix for that.

Do you have some more timings for your experiment?

**Comment by safesparrow** [comment]
Link to the BDN test I created:
https://github.com/safesparrow/fantomas/blob/parallel_cli/src/Fantomas.CLI.Benchmarks/Program.fs#L9

The test checks the FSharp compiler repository, using a revision from yesterday.

To run:
1. Change the local repo directory in 
https://github.com/safesparrow/fantomas/blob/97d54bc5fe208a08d15fb01b38818af55662ec0c/src/Fantomas.CLI.Benchmarks/Runners.fs#L13
2. Then:
```bash
cd src/Fantomas.CLI.Benchmarks
dotnet run -c Release
```

Results:
|                       Method |        Job | Server |     Mean | Error | Rank |         Gen0 |        Gen1 |       Gen2 | Allocated |
|----------------------------- |----------- |------- |---------:|------:|-----:|-------------:|------------:|-----------:|----------:|
| CheckSequentialWorkstationGc | Job-DOJKSB |  False | 14.635 s |    NA |    4 | 1126000.0000 | 152000.0000 | 32000.0000 |   8.55 GB |
| CheckInParallelWorkstationGc | Job-DOJKSB |  False | 10.486 s |    NA |    2 | 1107000.0000 | 211000.0000 | 11000.0000 |   8.55 GB |
|      CheckSequentialServerGc | Job-FVZCGD |   True | 12.602 s |    NA |    3 |   39000.0000 |  17000.0000 |  3000.0000 |   8.55 GB |
|      CheckInParallelServerGc | Job-FVZCGD |   True |  5.943 s |    NA |    1 |    8000.0000 |   3000.0000 |  1000.0000 |   8.55 GB |

I think more can be improved - it just needs some profiling work.

**Comment by nojaf** [comment]
Hey! This looks really promising. 
I would definitely be on board with this improvement!
My initial gut tells me this wouldn't need any flags and we can turn on server gc.
I might have some questions about what happens when one file fails but we can discuss that later.
Would you be interested in submitting a PR?

**Comment by safesparrow** [comment]
>I might have some questions about what happens when one file fails but we can discuss that later.

Yes, I haven't looked into that scenario.

There are also some shortcuts the PoC makes in how the logs are written (ie. there is no thread safety in the `printfn` statements).

>Would you be interested in submitting a PR?
Maybe at some point, but possibly not in the nearest future.

RE: Server GC
This might be worth enabling on its own if it's not controversial - should be just a matter of changing runtime config. One issue is whether it should be enabled for all 'modes', including the daemon mode.

**Comment by TheAngryByrd** [comment]
@nojaf Assign this issue to me
